src/traits.jl

```diff
@@ -51,6 +51,57 @@ end
     :($(ntuple(i -> Symbol(:dim_, i), N)))
 end
 
+Base.@pure function _dim_find(dimnames::NTuple{N,Symbol}, name::Symbol) where {N}
+    for i in 1:N
+        getfield(dimnames, i) === name && return i
+    end
+    return 0
+end
+
+"""
+    findaxis(img, name) -> axis
+
+Returns the axis (as in `axes(img, dim)`) that corresponds to the `name`. If no
+matching name is found any empy axis (i.e. `0:0`) is returned.
+"""
+@inline findaxis(A::T, name::Symbol) where {T} = _findaxis(namedaxes(A), name)
+function _findaxis(nt::NamedTuple{syms}, name::Symbol) where {syms}
+    if hasdim_noerror(syms, name)
+        return getfield(nt, name)
+    else
+        return 0:0
+    end
+end
+
+"""
+    finddim(img, name) -> Int
+
+Returns the dimension that has the corresponding `name`. If `name` doesn't
+match any of the dimension names `0` is returned. If `img` doesn't have `names`
+then the default set of names is searched (e.g., dim_1, dim_2, ...).
+"""
+@inline finddim(A::T, name::Symbol) where {T} = finddim(HasDimNames(T), A, name)
+finddim(::HasDimNames{false}, A::T, name::Symbol) where {T} = _dim_find(keys(namedaxes(A)), name)
+finddim(::HasDimNames{true}, A::T, name::Symbol) where {T} = _dim_find(names(A), name)
+
+
+"""
+    hasdim(x, name) -> Bool
+
+Returns `true` if `name` is present.
+"""
+hasdim(x::T, name::Symbol) where {T} = _hasdim(HasDimNames(T), x, name)
+_hasdim(::HasDimNames{true}, x::T, name::Symbol) where {T} = hasdim_noerror(names(x), name)
+_hasdim(::HasDimNames{false}, x::T, name::Symbol) where {T} = false
```

Comment:
`finddim(A, :dim_1)` could work even if `!hasdim(A, :dim_1)`. I worry a bit about API confusion/consistency in such cases. How important is it to have default names? Should we eliminate them and support operations only on explicitly-named axes?

Reply:
Good point. That needs to be more consistent. It seems kind of pointless to have it act on a default name that has no significant meaning. This makes me think that if there are default names they should probably be more meaningful, like those in AxisArrays. It may also help users if there's consistency with how AxisArrays names worked. I don't have a strong opinion either way, as long as I don't end being the guy that breaks JuliaImages.

Reply:
I'm not sure I'm a fan of `:row`, `:col`, and `:page` either. Those don't sound like image properties. At best they are linear algebra properties; just because we use 2d arrays for both doesn't mean image == matrix. (How often do you multiply an image times a vector??)

Can we just drop the default names altogether, and have these only operate on arrays with explicitly-assigned names?

Reply:
vote for `:dim_1`, `:dim_2`, and `:dim_3`